File: clojure/cars-assemble/src/cars_assemble.clj
(ns cars-assemble)

(def production-per-hour 221)

(defn production-rate
  "Returns the assembly line's production rate per hour,
   taking into account its success rate"
  [speed]
  (double
    (cond
      (= 0 speed) 0
      (and (>= speed 1) (<= speed 4)) (* speed production-per-hour)
      (and (>= speed 5) (<= speed 8)) (* 0.9 (* speed production-per-hour))
      (= speed 9) (* 0.8 (* speed production-per-hour))
      (= speed 10) (* 0.77 (* speed production-per-hour))
      ))
  )

(defn working-items
  "Calculates how many working cars are produced per minute"
  [speed]
  (int (Math/floor (/ (production-rate speed) 60)))
  )

File: go/cars-assemble/cars_assemble.go
package cars

import ( "math" )

// CalculateWorkingCarsPerHour calculates how many working cars are
// produced by the assembly line every hour.
func CalculateWorkingCarsPerHour(productionRate int, successRate float64) float64 {
	return float64(productionRate) * successRate / 100
}

// CalculateWorkingCarsPerMinute calculates how many working cars are
// produced by the assembly line every minute.
func CalculateWorkingCarsPerMinute(productionRate int, successRate float64) int {
	return int(math.Floor(float64(productionRate) * successRate / 100.0 / 60.0))
}

// CalculateCost works out the cost of producing the given number of cars.
func CalculateCost(carsCount int) uint {
	individual := carsCount % 10
	groups := carsCount / 10

	return uint(groups * 95000 + individual * 10000)
}
